File: ruby/src/trees/bst.rb
class BinarySearchTree
    attr_reader :left, :right, :parent, :count
    attr :value

    def initialize(value, parent)
        @left = nil
        @right = nil
        @parent = parent
        @value = value
        @count = 1
    end

    def insert(val)
        if val == @value
            @count += 1
        elsif val > @value
            if @right.nil?
                @right = BinarySearchTree.new(val, self)
            else
                @right.insert(val)
            end
        else
            if @left.nil?
                @left = BinarySearchTree.new(val, self)
            else
                @left.insert(val)
            end
        end
    end

    def remove(val)
        if val == @value
            # if this number's been inserted multiple times,
            # remove one instance
            if @count > 1
                @count -= 1
            elsif self.is_leaf?
                # just null the reference to this in our parent
                @parent.replace_child(@value, nil)
            elsif self.has_single_child?
                child = @left || @right

                @value = child.value
                @count = child.count

                child.remove(child.value)
            else
                # replace this with the greatest value in its left subtree
                child = @left.max

                # move the child to this node
                @value = child.value
                @count = child.count

                # get rid of the leftover object that we moved up, maintaining BST properties
                child.remove(child.value)
            end
        elsif val < @value
            # remove from left subtree
            if !@left.nil?
                @left.remove(val)
            end
        else
            # remove from right subtree
            if !@right.nil?
                @right.remove(val)
            end
        end
    end

    def contains?(val)
        return true if @value == val

        if val < @value
            return !@left.nil? && @left.contains?(val)
        else
            return !@right.nil? && @right.contains?(val)
        end
    end

    def is_leaf?()
        return @left.nil? && @right.nil?
    end

    def has_single_child?()
        return @left.nil? ^ @right.nil?
    end

    protected
        def replace_child(child_val, replacement_node)
            if child_val < value
                @left = replacement_node
            else
                @right = replacement_node
            end
        end

        def max
            return @right.max if !@right.nil? else return self
        end
end

File: c/trees/bst.c
#include <stdlib.h>
#include <stdio.h>

#include "bst.h"

void BinarySearchTree_init(BinarySearchTree *node, BinarySearchTree *parent, int n) {
    node->value = n;
    node->count = 1;
    node->parent = parent;
    node->left = NULL;
    node->right = NULL;
}

void BinarySearchTree_insert(BinarySearchTree *root, int n) {
    if(n == root->value) {
        // a node with the given value already exists -- increment it
        root->count += 1;
    } else if(n < root->value) {
        // left side insert
        if(root->left != NULL) {
            // insert into left subtree
            BinarySearchTree_insert(root->left, n);
        } else {
            // create left node
            root->left = malloc(sizeof(BinarySearchTree));
            BinarySearchTree_init(root->left, n);
        }
    } else {
        // right side insert
        if(root->right != NULL) {
            // insert into right subtree
            BinarySearchTree_insert(root->right, n);
        } else {
            // create right node
            root->right = malloc(sizeof(BinarySearchTree));
            BinarySearchTree_init(root->right, n);
        }
    }
}

void BinarySearchTree_remove(BinarySearchTree *root, int n) {
    if(root->value == n) {
        // delete this node

        if(root->count > 1) {
            // decrement the counter -- no need to delete it if there's multiple instances
            root->count -= 1;
        } else if(root->right == NULL && root->left == NULL) {
            // this is a leaf, just null the reference to it on its parent and free it

            if(root->parent->left == root) {
                root->parent->left = NULL;
            } else {
                root->parent->right = NULL;
            }

            free(root);
        } else {
            // single child -- replace this with the max of the left or the min of the right
            BinarySearchTree *node = NULL;

            // get min of right or max of left
            // default to max of left so it works for both single child nodes
            // as well as double child nodes
            if(root->left != NULL) {
                node = root->left;
                while(node->right != NULL) {
                    node = node->right;
                }
            } else {
                node = root->right;
                while(node->left != NULL) {
                    node = node->left;
                }
            }

            // set this node to be equal to the node we grabbed
            // then delete the old node
            root->value = node->value;
            root->count = node->count;

            // so it gets deleted when we remove it
            node->count = 1;

            BinarySearchTree_remove(node, node->value);
        }
    } else if(n < root->value) {
        // try to delete from left subtree

        if(root->left == NULL) {
            printf("Attempt to delete non-existing value %d from BST", n);
            return;
        }

        BinarySearchTree_remove(root->left, n);
    } else {
        // try to delete from right subtree

        if(root->right == NULL) {
            printf("Attempt to delete non-existing value %d from BST", n);
            return;
        }

        BinarySearchTree_remove(root->right, n);
    }
}

int BinarySearchTree_contains(BinarySearchTree *root, int n) {
    return (root->value == n) ||
                ((root->value < n && root->left != NULL && BinarySearchTree_contains(root->left)) ||
                ((root->value > n && root->right != NULL && BinarySearchTree_contains(root->right)))
            );
}

int BinarySearchTree_max(BinarySearchTree *root) {
    return root->right == NULL ? root->value ? BinarySearchTree_max(root->right);
}

int BinarySerachTree_min(BinarySearchTree *root) {
    return root->left == NULL ? root->value ? BinarySearchTree_max(root->left);
}
